### PR TITLE
Fixes deprecation warnings in `test_optim.py`

### LINKED ISSUE #67696: Warnings being produced when testing optim.
## 🐛 Bug

This is a small issue: pytorch tests have warnings when run on master (hash: 82f7f8d471241b0f417f932c59558774cc273041).

While this is clearly not a glaring issue, I think that tests ought not to produce warnings when running. It makes it harder for developers to see whether or not that is produced by their work.
Secondly, I think it is generally worth following the paradigms we expect to code with when we write tests.
Does this seem reasonable?

## To Reproduce

1. Install torch.
2. `cd pytorch`
3. `python test/test_optim.py`

Error messages:
```
/home/oliver/code/playground/pytorch-master/torch/optim/asgd.py:34: UserWarning: optimizer contains a parameter group with duplicate parameters; in future, this will cause an error; see github.com/pytorch/pytorch/issues/40967 for more information
  super(ASGD, self).__init__(params, defaults)
 ```
 This occurs for, as far as I can tell, every optimiser. As the error is the same, I will include only one copy.
 
 ```
 ........................../home/oliver/code/playground/pytorch-master/torch/optim/lr_scheduler.py:129: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  warnings.warn("Detected call of `lr_scheduler.step()` before `optimizer.step()`. "
```
Is produced several times too. Local testing makes me 99% sure this is caused in this snippet:
```
for _i in range(200):
            for scheduler in schedulers:
                if isinstance(scheduler, ReduceLROnPlateau):
                    val_loss = fn()
                    scheduler.step(val_loss)
                else:
                    scheduler.step()
            optimizer.step(fn)
```
[Source of this test code.](https://github.com/pytorch/pytorch/blob/82f7f8d471241b0f417f932c59558774cc273041/test/test_optim.py#L128)
  

## Expected behavior

Test should pass and issue no warnings.

## Environment

```
Collecting environment information...
PyTorch version: 1.11.0a0+gitcd51d2a
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~18.04) 9.4.0
Clang version: Could not collect # manual note, I built torch with `clang-6.0`
CMake version: version 3.19.6
Libc version: glibc-2.27

Python version: 3.8.12 (default, Oct 12 2021, 13:49:34)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-87-generic-x86_64-with-glibc2.17
Is CUDA available: False
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti
GPU 4: GeForce GTX 1080 Ti
GPU 5: GeForce GTX 1080 Ti
GPU 6: GeForce GTX 1080 Ti
GPU 7: GeForce GTX 1080 Ti

Nvidia driver version: 450.102.04
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.2
[pip3] torch==1.11.0a0+git82f7f8d
[conda] blas                      1.0                         mkl
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-include               2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0            py38h7f8727e_0
[conda] mkl_fft                   1.3.1            py38hd3c417c_0
[conda] mkl_random                1.2.2            py38h51133e4_0
[conda] numpy                     1.21.2           py38h20f2e39_0
[conda] numpy-base                1.21.2           py38h79a1101_0
[conda] torch                     1.11.0a0+git82f7f8d           dev_0    <develop>
```

## Additional context

One of the two issues is linked to #40967 -- a fix was made to emit a warning, but our tests were not updated.
I am happy to fix all of these issues, if that would be helpful!

cc @vincentqb @jbschlosser @albanD @mruberry

### PR DESCRIPTION
**Pull request by OliverFM**
Catches deprecation warnings when we call `scheduler.step(epoch)`
in tests.

Removes duplicate parameters to optimizers unless we are specifically
testing for that

Fixes #67696 

There is one warning remaining when I run this locally -- however that is due to the implementation of the `SequentialLR` Scheduler. I will open a new issue relating to that.

### CODE REVIEW

File: test/test_optim.py [hunk 1]
```diff
@@ -1039,7 +1039,8 @@ def _test_lr_is_constant_for_constant_epoch(self, scheduler):
 
         for _ in range(10):
             scheduler.optimizer.step()
-            scheduler.step(2)
+            with warnings.catch_warnings(record=True) as w:
```
Comment (albanD) [inline test/test_optim.py:1042]: Could you add a comment just before the catcher here to mention that we are hiding the warning due to the scheduler being used without an optimizer on purpose here.
Also to be extra sure that we don't hide anything else, you can add a `self.assertEqual(len(w), 1)` after the block to make sure we didn't catch any other unexpected warning.

Comment (OliverFM) [reply test/test_optim.py:1042]: Thanks for the suggestion -- will do!

Comment (OliverFM) [reply test/test_optim.py:1042]: Correct me if I am wrong, but the way I read this test:
we are using an optimizer, the reason we get a warning is because we are passing in the `epoch` parameter -- to test that `step(2)` is idempotent.

Comment (albanD) [reply test/test_optim.py:1042]: For this one looks like so.

File: test/test_optim.py [hunk 2]
```diff
@@ -2099,7 +2100,8 @@ def _test_with_epoch(self, schedulers, targets, epochs=10):
         optimizers = {scheduler.optimizer for scheduler in schedulers}
         for epoch in range(epochs):
             [optimizer.step() for optimizer in optimizers]
-            [scheduler.step(epoch) for scheduler in schedulers]  # step before assert: skip initial lr
+            with warnings.catch_warnings(record=True) as w:
```
Comment (albanD) [inline test/test_optim.py:2103]: Why is this one warning? we do the corresponding optimizer step above no?

File: test/test_optim.py [hunk 3]
```diff
@@ -2149,7 +2151,8 @@ def _test_against_closed_form(self, scheduler, closed_form_scheduler, epochs=10)
         targets = []
         for epoch in range(epochs):
             closed_form_scheduler.optimizer.step()
-            closed_form_scheduler.step(epoch)
+            with warnings.catch_warnings(record=True) as w:
```
Comment (albanD) [inline test/test_optim.py:2154]: Same this one should not warn?

**Review by albanD — COMMENTED** [review]
Sorry for the confused review. I am not 100% up to date with how and why all these warnings happen.

File: test/test_optim.py (hunk 3 above)
Comment (OliverFM) [reply test/test_optim.py:2154]: No worries about the confusion -- I relate to it too 😅 

In both cases the warning comes from passing in the epoch parameter -- which triggers this warning. https://github.com/pytorch/pytorch/blob/191b48b12f33e1e9525882da0c62b68686d69e42/torch/optim/lr_scheduler.py#L154

So I think that we ought to expect the warning to appear in all cases since we are making sure that the deprecated behaviour is functional for as long as it remains in the codebase

Comment (albanD) [reply test/test_optim.py:2154]: Ho ok! Sounds good! Thanks for checking.

**Review by albanD — APPROVED** [review]
Sounds good!

**Comment by OliverFM** [comment]
> Sounds good!

I have pushed the relevant fixes :)
